File: Classes/git/PBGitSVSubmoduleItem.m
//
//  PBGitSVSubmoduleItem.m
//  GitX
//
//  Created by Seth Raphael on 9/14/12.
//
//

#import "PBGitSVSubmoduleItem.h"
#import "PBGitBinary.h"
#import "PBEasyPipe.h"

@implementation PBSubmoduleInfo

+ (NSArray<PBSubmoduleInfo *> *)submodulesForRepositoryURL:(NSURL *)repositoryURL {
    NSMutableArray *submodules = [NSMutableArray array];
    
    NSString *gitPath = [PBGitBinary path];
    if (!gitPath) {
        gitPath = @"/usr/bin/git"; // Fallback
    }
    
    int exitCode = 0;
    NSString *output = [PBEasyPipe outputForCommand:gitPath
                                          withArgs:@[@"config", @"--file", @".gitmodules", @"--get-regexp", @"^submodule\\..*\\.path$"]
                                             inDir:[repositoryURL path]
                                          retValue:&exitCode];
    
    if (exitCode == 0 && output) {
        
        NSArray *lines = [output componentsSeparatedByString:@"\n"];
        for (NSString *line in lines) {
            if ([line length] > 0) {
                NSRange pathRange = [line rangeOfString:@"submodule."];
                NSRange dotPathRange = [line rangeOfString:@".path "];
                
                if (pathRange.location != NSNotFound && dotPathRange.location != NSNotFound) {
                    NSString *name = [line substringWithRange:NSMakeRange(pathRange.location + pathRange.length, 
                                                                         dotPathRange.location - (pathRange.location + pathRange.length))];
                    NSString *path = [[line substringFromIndex:dotPathRange.location + dotPathRange.length] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
                    
                    PBSubmoduleInfo *info = [[PBSubmoduleInfo alloc] init];
                    info.name = name;
                    info.path = path;
                    info.parentRepositoryURL = repositoryURL;
                    [submodules addObject:info];
                }
            }
        }
    }
    
    return [submodules copy];
}

@end

@implementation PBGitSVSubmoduleItem

+ (id) itemWithSubmodule:(PBSubmoduleInfo*)submodule
{
    PBGitSVSubmoduleItem* item = [[self alloc] init];
	item.submodule = submodule;
    return item;
}

- (NSString *)title
{
	return self.submodule.name;
}

- (NSURL *)path
{
	NSURL *parentURL = self.submodule.parentRepositoryURL;
	NSURL *result = [parentURL URLByAppendingPathComponent:self.submodule.path];
	return result;
}
@end

File: Classes/git/PBGitSVSubmoduleItem.swift
import Foundation

@objc(PBSubmoduleInfo)
@objcMembers
final class PBSubmoduleInfo: NSObject {
    var name: String?
    var path: String?
    var parentRepositoryURL: URL?

    @objc(submodulesForRepositoryURL:)
    class func submodules(for repositoryURL: URL) -> [PBSubmoduleInfo] {
        let gitPath = PBGitBinary.path() ?? "/usr/bin/git"
        var exitCode: Int32 = 0
        guard let output = PBEasyPipe.outputForCommand(
            gitPath,
            withArgs: [
                "config",
                "--file",
                ".gitmodules",
                "--get-regexp",
                "^submodule\\..*\\.path$"
            ],
            inDir: repositoryURL.path,
            retValue: &exitCode
        ), exitCode == 0 else {
            return []
        }

        var submodules: [PBSubmoduleInfo] = []

        output.split(separator: "\n").forEach { rawLine in
            guard !rawLine.isEmpty else { return }
            let line = String(rawLine)

            guard let pathRange = line.range(of: "submodule."),
                  let dotPathRange = line.range(of: ".path "),
                  pathRange.upperBound <= dotPathRange.lowerBound else {
                return
            }

            let name = String(line[pathRange.upperBound..<dotPathRange.lowerBound])
            let pathComponent = String(line[dotPathRange.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
            guard !name.isEmpty, !pathComponent.isEmpty else {
                return
            }

            let info = PBSubmoduleInfo()
            info.name = name
            info.path = pathComponent
            info.parentRepositoryURL = repositoryURL
            submodules.append(info)
        }

        return submodules
    }
}

@objc(PBGitSVSubmoduleItem)
@objcMembers
final class PBGitSVSubmoduleItem: PBSourceViewItem {
    var submodule: PBSubmoduleInfo?

    @objc(itemWithSubmodule:)
    class func item(with submodule: PBSubmoduleInfo) -> PBGitSVSubmoduleItem {
        let item = PBGitSVSubmoduleItem()
        item.submodule = submodule
        return item
    }

    override var title: String! {
        get { submodule?.name }
        set { super.title = newValue }
    }

    var path: URL? {
        guard let parent = submodule?.parentRepositoryURL,
              let relative = submodule?.path else {
            return nil
        }

        return parent.appendingPathComponent(relative)
    }
}
